File: src/main/java/com/gitlab/daring/image/command/feature/ShowFeatureMatchesCommand.java
package com.gitlab.daring.image.command.feature;

import com.gitlab.daring.image.command.BaseCommand;
import com.gitlab.daring.image.command.CommandEnv;
import com.gitlab.daring.image.command.parameter.IntParam;
import com.gitlab.daring.image.command.parameter.StringParam;
import com.gitlab.daring.image.features.DMatchResult;
import one.util.streamex.StreamEx;
import org.bytedeco.javacpp.opencv_core.DMatch;
import org.bytedeco.javacpp.opencv_core.DMatchVector;
import org.bytedeco.javacpp.opencv_core.Mat;
import org.bytedeco.javacpp.opencv_core.Scalar;

import java.nio.ByteBuffer;

import static com.gitlab.daring.image.command.CommandScriptUtils.runCommand;
import static com.gitlab.daring.image.util.OpencvConverters.toJava;
import static org.bytedeco.javacpp.opencv_features2d.DrawMatchesFlags.NOT_DRAW_SINGLE_POINTS;
import static org.bytedeco.javacpp.opencv_features2d.drawMatches;

public class ShowFeatureMatchesCommand extends BaseCommand {

    final IntParam maxFeatures = intParam(100, "0-100");
    final StringParam title = stringParam("matches");

    final Scalar color = Scalar.all(-1);
    final ByteBuffer mask = ByteBuffer.allocate(0);
    final Mat rm = new Mat();

    public ShowFeatureMatchesCommand(String... args) {
        super(args);
    }

    @Override
    public void execute(CommandEnv env) {
        DMatchResult mr = env.matchResult;
        DMatch[] matches = StreamEx.of(toJava(mr.matches)).limit(maxFeatures.v).toArray(DMatch.class);
        drawMatches(
            mr.points1.mat, mr.points1.points,
            mr.points2.mat, mr.points2.points,
            new DMatchVector(matches), rm,
            color, color, mask, NOT_DRAW_SINGLE_POINTS
        );
        env.mat = rm;
        runCommand(env, "show", title.v);
    }
}

File: src/main/kotlin/com/gitlab/daring/image/command/feature/ShowFeatureMatchesCommand.kt
package com.gitlab.daring.image.command.feature

import com.gitlab.daring.image.command.CommandEnv
import com.gitlab.daring.image.command.CommandScriptUtils.runCommand
import com.gitlab.daring.image.command.KBaseCommand
import com.gitlab.daring.image.util.OpencvConverters.toJava
import org.bytedeco.javacpp.opencv_core
import org.bytedeco.javacpp.opencv_core.DMatchVector
import org.bytedeco.javacpp.opencv_features2d.DrawMatchesFlags.NOT_DRAW_SINGLE_POINTS
import org.bytedeco.javacpp.opencv_features2d.drawMatches
import java.nio.ByteBuffer

internal class ShowFeatureMatchesCommand(args: Array<String>) : KBaseCommand(args) {

    val maxFeatures = intParam(100, "0-100")
    val title = stringParam("matches")

    val color = opencv_core.Scalar.all(-1.0)
    val mask = ByteBuffer.allocate(0)
    val rm = opencv_core.Mat()

    override fun execute(env: CommandEnv) {
        val (ps1, ps2, ms) = env.matchResult
        val matches = toJava(ms).take(maxFeatures.v).toTypedArray()
        drawMatches(
                ps1.mat, ps1.points,
                ps2.mat, ps2.points,
                DMatchVector(*matches), rm,
                color, color, mask, NOT_DRAW_SINGLE_POINTS
        )
        env.mat = rm
        runCommand(env, "show", title.v)
    }

}